Key renew

fatih opened the pull request:
This is the second part of multi key based kontrol. It includes an automatic key checker and replacing old keys with new keys, if the old key is expired.

--

* We test for Go `1.4` and `tip` now. `1.3` is removed as we built the binaries already with `1.4` and it's better to look forwards
* Fix many race conditions in the client.go code which was detected with the race detector.
* `register` method now returns a new key if you try to register with an old key. 
* A new `getKey` method is added to Kontrol to retrieve a new public key if the current one is invalidated. 
* Fix internal memory key pair storage, which was failing if public key was empty
* Last set public and private keys is now a list of keys. Because we might invalidate a key, so once invalidated we need use the last set key, for that we we need to store all keys to have a reference to it.
* An external `DeleteKeyPair` method is added which deleted keys
* Kite can now retrieve and update the public key in memory with `GetKey()` (single usage) or the `NewKeyRenewer()` which updates the key every given internal duration.

cihangir commented:
> We test for Go 1.4 and tip now. 1.3 is removed as we built the binaries already with 1.4 and it's better to look forwards

We can still support 1.3 with no additional effort.

cihangir commented on kontrol/kontrol.go:
```diff
@@ -128,6 +131,24 @@ func (k *Kontrol) AddAuthenticator(keyType string, fn func(*kite.Request) error)
+		Public: public,
+	})
+
+	// delete latest added key so it doesn't get picked up
```
how do you make sure that given id/public key pair is the latest one?

fatih replied:
Because we use a slice and the keys are appended each time we use `AddKeyPair`:

```go
	// set last set key pair
	k.lastPublic = append(k.lastPublic, public)
	k.lastPrivate = append(k.lastPrivate, private)
```

cihangir commented on kontrol/postgres.go:
```diff
@@ -415,6 +416,36 @@ func (p *Postgres) DeleteKey(keyPair *KeyPair) error {
 }
 
+func (p *Postgres) IsValid(public string) error {
+	psql := sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
```
you could utilze `GetKeyFromPublic`

fatih replied:
I've did explicitly so validity can be checked differently. For example we might check it via `deleted_at` field, someone else could made other checks. Just using `Get...` methods would be limited.

cihangir commented on kontrol/keypair.go:
```diff
@@ -47,6 +47,10 @@ type KeyPairStorage interface {
+
+	// Is valid checks if the given publicKey is valid or not. It's up to the
+	// implementer how to implement it. A valid public key returns a nil error.
+	IsValid(publicKey string) error
```
same functionality can be handled with `GetKeyFromPublic` function if it returns and exported `ErrNotFound` error.

cihangir commented on kontrolclient.go:
```diff
@@ -179,6 +179,43 @@ func (k *Kite) GetToken(kite *protocol.Kite) (string, error) {
+// NewKeyRenewer renews the internal key every given interval
+func (k *Kite) NewKeyRenewer(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	for range ticker.C {
```
better to close ticker when we call kite.Close

cihangir commented:
Review status: 0 of 12 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[kontrol/kontrol.go, line 146 \[r1\]](https://reviewable.io:443/reviews/koding/kite/115#-JrurBHVjUB2kalR-0pw-r1-146)** ([raw file](https://github.com/koding/kite/blob/4fb91074e7367c62990464e6e146c3d077514cc2/kontrol/kontrol.go#L146)):</sup>
Ok but this doesn't prevent me from calling delete for any item that is currently in the middle, what i am missing here?

---

<sup>**[kontrol/postgres.go, line 420 \[r1\]](https://reviewable.io:443/reviews/koding/kite/115#-JrurBHVjUB2kalR-0pz-r1-420)** ([raw file](https://github.com/koding/kite/blob/4fb91074e7367c62990464e6e146c3d077514cc2/kontrol/postgres.go#L420)):</sup>
if someone would check differently, it means that, they need to implement our interface, with the same approach they can implement their own logic in regarding function, what do you think?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/koding/kite/115)
<!-- Sent from Reviewable.io -->

fatih commented:
>Ok but this doesn't prevent me from calling delete for any item that is currently in the middle, what i am missing here?

I've fixed it.

> if someone would check differently, it means that, they need to implement our interface, with the same approach they can implement their own logic in regarding function, what do you think?

The `Get` functions are returning very limited information. It's only the `id`, `public` and `private` key. So based on those, implementing a validation is really hard. I know that `Get` can return a `not found` error, however does a `not found` really means the key is not valid? Maybe the keypair storage has connectivity problem to the DB or it's something totally different. It's just `IsValid` just makes it simpler to add custom logic. 

> We can still support 1.3 with no additional effort.

I've fixed it. It's reverted back to use `1.3` and `1.4`

cihangir commented:
:+1:
